src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+                }
+            }
+
+            if (quals.length % 2 != 0) {
+                quals[quals.length / 2] -= 33;
+            }
+
+            return quals;
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private int getMateAlignmentStartImpl() {
+        try {
+            if (isAligned && getReadPairedFlag() && !getMateUnmappedFlag()) {
+                Alignment mate = getCurrentAlignment().getMateAlignment();
+                return (int) mate.getAlignmentPosition() + 1;
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+        return 0;
+    }
+
+    private String getMateReferenceNameImpl() {
+        try {
+            if (isAligned && getReadPairedFlag() && !getMateUnmappedFlag()) {
+                return getCurrentAlignment().getMateReferenceSpec();
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+        return SAMRecord.NO_ALIGNMENT_REFERENCE_NAME;
+    }
+
+    private int getInferredInsertSizeImpl() {
+        try {
+            if (isAligned) {
+                return (int) getCurrentAlignment().getTemplateLength();
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+        return 0;
+    }
+
+    // ===== flags =====
+
+    private boolean getReadNegativeStrandFlagImpl() {
+        try {
+            if (isAligned) {
+                return getCurrentAlignment().getIsReversedOrientation();
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+        return false;
+    }
+
+    private boolean getReadPairedFlagImpl() {
+        try {
+            if (isAligned) {
+                return getCurrentAlignment().isPaired();
+            } else {
+                return getCurrentUnalignedRead().getNumFragments() > 1;
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean getNotPrimaryAlignmentFlagImpl() {
+        try {
+            if (isAligned) {
+                return getCurrentAlignment().getAlignmentCategory() == Alignment.secondaryAlignment;
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        return false;
+    }
+
+    private boolean getMateNegativeStrandFlagImpl() {
+        try {
+            if (isAligned && getReadPairedFlag() && !getMateUnmappedFlag()) {
+                Alignment mate = getCurrentAlignment().getMateAlignment();
+                return mate.getIsReversedOrientation();
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        return false;
+    }
+
+    private boolean getMateUnmappedFlagImpl() {
+        try {
+            if (isAligned) {
+                return !getCurrentAlignment().hasMate();
```

Comment:
There actually is. Alignment.hasMate() returns true if it has aligned mate, but Alignment.isPaired() checks if any mate exists (both mapped and unmapped)